docs/Client.md

```diff
@@ -0,0 +1,883 @@
+
+      client.close()
+      fastifyServer.close()
+      nodeServer.close()
+    })
+  })
+})
+```
+
+### `Client.upgrade(options[, callback])`
+
+Upgrade the client to a different protocol.
+
+Arguments:
+
+* **options** `UpgradeOptions`
+  
+* **callback** `(error: Error | null, data: UpgradeData) => void` (optional) - If a `callback` is not provided, a `Promise` is returned
+
+Returns: `void | Promise<UpgradeData>` - Only returns a `Promise` if no `callback` argument was passed
+
+#### Parameter: `UpgradeOptions`
+
+* **path** `string`
+* **method** `string` (optional) - Default: `'GET'`
+* **headers** `UndiciHeaders` (optional) - Default: `null`
+* **protocol** `string` (optional) - Default: `'Websocket'` - A string of comma separated protocols, in descending preference order.
+* **signal** `AbortSignal | EventEmitter | null` (optional) - Default: `null`
+
+#### Parameter: `UpgradeData`
+
+* **headers** `http.IncomingHeaders`
+* **socket** `stream.Duplex`
+* **opaque** `unknown`
+
+#### Example 1 - Basic Upgrade Request
+
+```js
+'use strict'
+const { Client } = require('undici')
+const { createServer } = require('http')
+
+const server = createServer((request, response) => {
+  response.statusCode = 101
+  response.setHeader('connection', 'upgrade')
+  response.setHeader('upgrade', request.headers.upgrade)
+  response.end()
+})
+
+server.listen(() => {
+  const client = new Client(`http://localhost:${server.address().port}`)
+
+  client
+    .upgrade({ path: '/' })
+    .then(({ headers, socket }) => {
+      socket.on('end', () => {
+        console.log(`upgrade: ${headers.upgrade}`)
+        client.close()
+        server.close()
+      })
+      socket.end()
+    })
+    .catch(error => {
+      console.error(error)
+      client.close()
+      server.close()
+    })
+})
+```
+
+## Instance Properties
+
+### `Client.busy`
+
+* `boolean`
+
+`true` if pipeline is saturated or blocked. Indicates whether dispatching further requests is meaningful.
+
+### `Client.closed`
+
+* `boolean`
+
+`true` after `client.close()` has been called.
+
+### `Client.connected`
+
+* `boolean`
+
+`true` if the client has an active connection. The client will lazily create a connection when it receives a request and will destroy it if there is no activity for the duration of the `timeout` value.
+
+### `Client.destroyed`
+
+* `boolean`
+
+`true` after `client.destroyed()` has been called or `client.close()` has been called and the client shutdown has completed.
+
+### `Client.pending`
+
+* `number`
+
+Number of queued requests.
+
+### `Client.pipelining`
+
+* `number`
+
+Property to get and set the pipelining factor.
+
+### `Client.running`
+
+* `number`
+
+Number of inflight requests.
+
+### `Client.size`
+
+* `number`
+
+Number of pending and running requests.
+
+## Instance Events
+
+### Event: `'connect'`
+
+Emitted when a socket has been created and connected. The client will connect once `Client.size > 0`.
+
+### Event: `'disconnect'`
+
+Returns:
```

Comment:
returns?

Reply:
Yeah not sure how to best denote events. They are also missing examples I will add now

Reply:
Parameters? events are basically the same as function invocations.

Reply:
Okay I found two disconnect event instances:

https://github.com/nodejs/undici/blob/10486556f3a2c40a7e90851da7867d1868fb0027/lib/core/client.js#L493

https://github.com/nodejs/undici/blob/10486556f3a2c40a7e90851da7867d1868fb0027/lib/core/client.js#L823

One calls with `client, error` and the other just `error`. Is this okay? Should we fix the inconsistency and also return client in the later disconnect event?

Reply:
The first one is a bug. Would you mind creating a PR for it (and any other instances)?